README.md
# Timer for Harvest
[Harvest](https://www.getharvest.com/) client implemented using
[GTK](https://www.gtk.org/) and [Rust](https://www.rust-lang.org/) for Linux
and the various BSD's.

## Screenshots

![Main window](/assets/main-window.png?raw=true "The main window")
![Popup](/assets/popup.png?raw=true "The time entry popup")

## Installation
On the Timer for Harvest github page, click on the 
[releases link](https://github.com/frenkel/timer-for-harvest/releases). The 
newest release is listed at the top of the page. It contains pre-build 
binaries for Ubuntu 20.04+ and Fedora 34+, as well as other distributions. 
You will be notified of new releases by a message in the main window.

## Usage
After installation your GNOME shell should be able to find the
application when you type "Timer for Harvest" in the activity searcher.

### First usage
Upon first launch you will first see a web browser start. This will open an
authorization flow from Harvest. When you've completed these steps you will
end up on a white page with the text "Authorized successfully." The actual
application will start now.

### Daily usage
Just like the Harvest web interface there are some handy keyboard shortcuts:
- **F5** in the main window will refresh the time entries list. This can be
  usefull when you updated the entries using a different interface and
  Timer for Harvest still shows the old state.
- **N** in the main window opens the new time entry popup.
- **Esc** closes the time entry popup.
- **Enter** activates the "Save Timer" button in the time entry popup.

## Security
Username and password details are never seen by Timer for Harvest. A web
browser is used to authorize Timer for Harvest access to your account. This
authorization is stored in the form of an authorization token, which Harvest
lets expire in 14 days. Leaking this token would thus give somebody access to
your account for a maximum of 14 days.

The authorization token is currently stored on the file system, namely in
$XDG\_CONFIG\_HOME/timer-for-harvest.json. In the future we hope to move this
token to a more secure location, such as
[libsecret](https://wiki.gnome.org/Projects/Libsecret).

## Wishlist
- Idle detection to ask user whether idle time should be subtracted or booked
  as a new time entry. For example, using this dbus call when using GNOME:
  `dbus-send --print-reply --dest=org.gnome.Mutter.IdleMonitor /org/gnome/Mutter/IdleMonitor/Core org.gnome.Mutter.IdleMonitor.GetIdletime`
- Authentication token storage in [libsecret](https://wiki.gnome.org/Projects/Libsecret).
- [Improve UI](https://github.com/frenkel/timer-for-harvest/issues/34) to speed up the new entry process.

## Building
If you want to build the application yourself you'll need rust, cargo, bindgen, clang and the gtk3
development libraries installed. You can then run `cargo build --release` to generate
the binary in `target/release/`.

## Uninstall

To ininstall, simply run the following command in a terminal based on your distro: 

- Fedora / RHEL or other RPM-based distros : `sudo dnf remove timer-for-harvest`
- Debian / Ubuntu / Linux Mint or other apt-based distros: `sudo apt remove timer-for-harvest`

## Donations
If you like the software I create, please consider donating through
[Paypal](https://paypal.me/frankgroeneveld) or
[Github Sponsors](https://github.com/sponsors/frenkel).


CHANGELOG.md
# Changelog

## [UNRELEASED] - YYYY-MM-DD

## [0.3.10] - 2022-08-01

- Build against OpenSSL 1.1 for Ubuntu 20.04.

## [0.3.10] - 2022-07-29

- Build against OpenSSL 3 for Ubuntu 22.04. 

## [0.3.9] - 2022-06-11

- Updates to dependencies. These dependencies needed to be updated because of various security issues.
- Clear selected task when switching project. Previously an invalid entry was still selected, which stopped the save button from working.

## [0.3.8] - 2021-09-21

- Re-enable the ability to resize the main window, implemented by @mlunax.
- The license of Timer for Harvest will no longer show up as proprietary in GNOME Software.

## [0.3.7] - 2021-08-26

- Show correct button label when adding hours instead of starting a new timer, implemented by @mlunax.
- Fixed a crash on resume-from-suspend because the time entries could not load while network was not up yet.

## [0.3.6] - 2021-03-12

This release includes escaping for special characters such as & in task, project en client names. Thanks @wvengen for reporting this bug.
Furthermore the update link is now clickable thanks to @iatanas0v.

## [0.3.5] - 2021-02-26

Thanks to @KillerCodeMonkey we have another release! This time he contributed changes to the main window that keep it a static size with a scrollbar. Thanks again @KillerCodeMonkey!

## [0.3.4] - 2021-02-05

Thanks to @KillerCodeMonkey we now have multi-line description support!
We also had to switch to a different client\_id for the authorization at Harvest. You shouldn't notice this but we're reporting this so that you know it is not a security incident.

## [0.3.3] - 2020-10-23

This release includes the following new features and bugfixes:

- Auto refresh when receiving window focus, idea by @Flimm
- Add client name to project chooser, requested by @jarnalyrkar
- New entry's are now added on the chosen date instead of the current date
- Edit button is now re-enabled when cancelling the edit popup


## [0.3.2] - 2020-08-21

This release consist of a major architectural rewrite. As an end user this shouldn't be noticeable, except for some interface alignment changes. However one user noticeable feature was added: the application will now check whether a new version is available and show a notification when one is available. This works by checking a DNS TXT record, so we cannot misuse this to track usage of the application.

## [0.3.1] - 2020-06-26

This release adds two new features contributed by @jaspervandenberg:

- Add a button to go to the current date
- Ask for confirmation before deleting an entry

## [0.3.0] - 2020-06-09

This release includes various interface improvements and one big new feature: the ability to switch to different days.

## [0.2.1] - 2020-04-14

This is a very small release that fixes one important bug: the project code can be null, which would result in a crash.

## [0.2.0] - 2020-04-08

This release improves the user experience a lot. All API communication has been moved to a background thread, resulting in a more response user interface. Furthermore, a number of error messages have been approved, which should make debugging problems easier.

## [0.1.12] - 2020-02-19

The main window acquired a lot of white spacing after enabling truncate on the labels. This release reduces it drastically.

## [0.1.11] - 2020-02-18

This release mostly documents the application, but it also fixes a line wrapping problem for long project names

## [0.1.10] - 2020-02-11

This release loads project data on startup instead of everytime you click a button. This makes using it on a daily basis a lot faster, at the expense of a slower startup time.

## [0.1.9] - 2020-01-14

This release is the first release with packages for Fedora 31 and Ubuntu 19.10. A small bug that would occur on Ubuntu when authorizing was fixed in this release as well.

## [0.1.8] - 2019-12-31

This release adds the ability to save and load the previously obtained OAuth tokens to disk. This means that the authorization flow needs to be run only once every two weeks, making it a lot more user-friendly.

## [0.1.7] - 2019-12-29

This is the first user-friendly release because it replaces the usage of developer tokens with an OAuth authorization flow. Currently the obtained authorization is not remembered between different application launches, but support for that will follow in a later release.

## [0.1.6] - 2019-12-27

Improvements in this release:

- Re-order time entries of the day correctly, newest at the bottom
- Don't require a terminal when running, by removing println!() statements

## [0.1.5] - 2019-12-26

This release includes some important refactoring, which allowed us to add the ability for running timers to update their labels.

## [0.1.4] - 2019-12-24

This release makes usage a lot faster by adding the following features:

- Auto completion in the project chooser combobox
- Auto completion in the task chooser combobox
- Close timer popup without saving by pressing escape
- Open timer popup by pressing n

## [0.1.3] - 2019-12-10

This release adds some more polish to the whole application. Notable changes include:

- Wrap notes in main window.
- Don't crash when time entry has no notes.
- Pressing enter in popup now defaults to saving the timer.
- Task loading has been modified to don't rely on admin-permission API calls. This change also allowed for an extra API call to be removed, resulting in a nicer workflow.

## [0.1.2] - 2019-12-03

This release adds some small but important improvements.

- Implement workaround for updating running timers without overwriting hours
- Show more time entry info, add some style improvements
- Order projects case-insensitive
- Don't crash after switching project where task is missing

## [0.1.1] - 2019-11-28

This release fixes various bugs:

- Fix bug in time formatting where 1 minute would show as 10
- Don't crash when editing timers without notes
- Refresh time entries list after edits

## [0.1.0] - 2019-11-26


tests/lib.rs
#[cfg(test)]
mod test {
    #[test]
    fn should_convert_duration_correctly() {
        assert_eq!("1:00", timer_for_harvest::f32_to_duration_str(1.0));
        assert_eq!("0:01", timer_for_harvest::f32_to_duration_str(1.0 / 60.0));
        assert_eq!("0:05", timer_for_harvest::f32_to_duration_str(5.0 / 60.0));
        assert_eq!("0:10", timer_for_harvest::f32_to_duration_str(10.0 / 60.0));
        assert_eq!("1:00", timer_for_harvest::f32_to_duration_str(59.9 / 60.0));
    }

    #[test]
    fn should_not_crash_duration_str_to_f32() {
        assert_eq!(0.0, timer_for_harvest::duration_str_to_f32("0:00"));
        assert_eq!(1.5, timer_for_harvest::duration_str_to_f32("1:30"));
        assert_eq!(1.0, timer_for_harvest::duration_str_to_f32("1"));
    }

    #[test]
    fn should_parse_account_id() {
        assert_eq!(
            "123",
            timer_for_harvest::parse_account_details("GET /?access_token=abc&scope=harvest%3A123")
                .1
        );
        assert_eq!(
            "123",
            timer_for_harvest::parse_account_details("GET /?scope=harvest%3A123&access_token=abc")
                .1
        );
    }

    #[test]
    fn should_parse_access_token() {
        assert_eq!(
            "abc",
            timer_for_harvest::parse_account_details("GET /?access_token=abc&scope=harvest%3A123")
                .0
        );
        assert_eq!(
            "abc",
            timer_for_harvest::parse_account_details("GET /?scope=harvest%3A123&access_token=abc")
                .0
        );
    }

    #[test]
    fn should_parse_expires_in() {
        assert_eq!(
            "123",
            timer_for_harvest::parse_account_details("GET /?expires_in=123&scope=harvest%3A456").2
        );
        assert_eq!(
            "123",
            timer_for_harvest::parse_account_details("GET /?scope=harvest%3A456&expires_in=123").2
        );
    }

    #[test]
    fn should_format_timeentry_notes_for_list() {
        assert_eq!("Lorem Ipsum - 1234567890 - 1234567890 - 1234567890 - 1234567890 - 1234567890 - 1...", timer_for_harvest::format_timeentry_notes_for_list(&"Lorem Ipsum\n\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890", None));
        assert_eq!("Lorem Ipsum - 1...", timer_for_harvest::format_timeentry_notes_for_list(&"Lorem Ipsum\n\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890\n1234567890", Some(15)));
    }
}


src/main.rs
mod app;
mod popup;
mod ui;

use app::App;
use std::env::args;
use std::sync::mpsc;
use timer_for_harvest::Harvest;
use ui::Ui;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = args().collect();

    if args.len() == 2 && &args[1] == "--version" {
        println!("{}", Harvest::user_agent());
    } else {
        let (to_ui, from_app) = glib::MainContext::channel(glib::PRIORITY_DEFAULT);
        let (to_app, from_ui) = mpsc::channel();

        let api = Harvest::new(args.get(1));
        match api {
            None => return Ok(()),
            Some(harvest) => {
                let app = App::new(to_ui, harvest);
                let ui = Ui::new(to_app);

                App::handle_ui_signals(app, from_ui);
                Ui::handle_app_signals(ui, from_app);
            }
        }
    }

    Ok(())
}


src/background.rs
use crate::ui;
use std::sync::mpsc;
use timer_for_harvest::*;

pub enum Event {
    RetrievedProjectAssignments(Vec<ProjectAssignment>),
    RetrievedTimeEntries(Vec<TimeEntry>),
    TimerStarted,
    TimerStopped,
    TimerRestarted,
    TimerUpdated,
    TimerDeleted,
    Loading(Option<u32>),
    OpenPopup(u32), /* actually sent from ui to itself */
    SetTitle(String),
}

pub fn handle_events(
    api: &Harvest,
    from_foreground: mpsc::Receiver<ui::Event>,
    to_foreground: &glib::Sender<Event>,
) {
    let mut for_date = chrono::Local::today().naive_local();

    for event in from_foreground {
        match event {
            ui::Event::RetrieveProjectAssignments => {
                to_foreground
                    .send(Event::Loading(None))
                    .expect("Sending message to foreground");

                println!("Retrieving project assignments");
                let mut project_assignments = api.active_project_assignments();
                project_assignments.sort_by(|a, b| {
                    a.project
                        .name
                        .to_lowercase()
                        .cmp(&b.project.name.to_lowercase())
                });
                to_foreground
                    .send(Event::RetrievedProjectAssignments(project_assignments))
                    .expect("Sending message to foreground");
            }
            ui::Event::RetrieveTimeEntries => {
                retrieve_time_entries(&to_foreground, &api, &for_date);
            }
            ui::Event::StartTimer(project_id, task_id, notes, hours) => {
                to_foreground
                    .send(Event::Loading(None))
                    .expect("Sending message to foreground");

                println!("Starting timer");
                api.start_timer(project_id, task_id, &notes, hours);
                to_foreground
                    .send(Event::TimerStarted)
                    .expect("Sending message to foreground");
            }
            ui::Event::StopTimer(id) => {
                to_foreground
                    .send(Event::Loading(Some(id)))
                    .expect("Sending message to foreground");

                println!("Stopping timer");
                api.stop_timer(id);
                to_foreground
                    .send(Event::TimerStopped)
                    .expect("Sending message to foreground");
            }
            ui::Event::RestartTimer(id) => {
                to_foreground
                    .send(Event::Loading(Some(id)))
                    .expect("Sending message to foreground");

                println!("Restarting timer");
                api.restart_timer(id);
                to_foreground
                    .send(Event::TimerRestarted)
                    .expect("Sending message to foreground");
            }
            ui::Event::UpdateTimer(
                id,
                project_id,
                task_id,
                notes,
                hours,
                is_running,
                spent_date,
            ) => {
                to_foreground
                    .send(Event::Loading(Some(id)))
                    .expect("Sending message to foreground");

                println!("Updating timer");
                api.update_timer(
                    id, project_id, task_id, notes, hours, is_running, spent_date,
                );
                to_foreground
                    .send(Event::TimerUpdated)
                    .expect("Sending message to foreground");
            }
            ui::Event::DeleteTimer(id) => {
                to_foreground
                    .send(Event::Loading(Some(id)))
                    .expect("Sending message to foreground");

                println!("Deleting timer");
                api.delete_timer(id);
                to_foreground
                    .send(Event::TimerDeleted)
                    .expect("Sending message to foreground");
            }
            ui::Event::PreviousDate => {
                for_date = for_date.pred();
                retrieve_time_entries(&to_foreground, &api, &for_date);
            }
            ui::Event::NextDate => {
                for_date = for_date.succ();
                retrieve_time_entries(&to_foreground, &api, &for_date);
            }
        }
    }
}

fn retrieve_time_entries(
    to_foreground: &glib::Sender<Event>,
    api: &Harvest,
    for_date: &chrono::NaiveDate,
) {
    to_foreground
        .send(Event::Loading(None))
        .expect("Sending message to foreground");

    println!("Retrieving time entries for {}", for_date);
    let user = api.current_user();
    to_foreground
        .send(Event::RetrievedTimeEntries(api.time_entries_for(
            user,
            for_date.to_string(),
            for_date.to_string(),
        )))
        .expect("Sending message to foreground");

    let title = format!("Harvest - {}", for_date.format("%a %-d %b"));
    to_foreground
        .send(Event::SetTitle(title))
        .expect("Sending message to foreground");
}


src/lib.rs
use dirs;
use hyper;
use serde;
use serde_json::json;
use std::fs::write;
use std::fs::File;
use std::io::Read;
use std::io::Write;
use std::net::TcpListener;
use std::net::TcpStream;
use std::path::PathBuf;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Harvest {
    token: String,
    account_id: u32,
    expires_at: u64,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct Project {
    pub id: u32,
    pub name: String,
    pub code: Option<String>,
    pub client: Option<Client>,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct ProjectAssignment {
    pub id: u32,
    pub project: Project,
    pub task_assignments: Vec<TaskAssignment>,
    pub client: Client,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct Client {
    pub id: u32,
    pub name: String,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct Task {
    pub id: u32,
    pub name: String,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct TaskAssignment {
    pub id: u32,
    pub task: Task,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct TimeEntry {
    pub id: u32,
    pub project: Project,
    pub client: Client,
    pub hours: f32,
    pub user: User,
    pub spent_date: String,
    pub task: Task,
    pub notes: Option<String>,
    pub is_running: bool,
}

/* a partially filled TimeEntry with id's instead of objects (Project etc) */
#[derive(serde::Serialize, serde::Deserialize)]
pub struct Timer {
    pub id: Option<u32>,
    pub project_id: u32,
    pub task_id: u32,
    pub spent_date: Option<String>,
    pub notes: Option<String>,
    pub hours: Option<f32>,
    pub is_running: bool,
}

/* a partially filled TimeEntry with id's instead of objects (Project etc) */
#[derive(serde::Serialize, serde::Deserialize)]
pub struct TimerWithoutHours {
    pub id: Option<u32>,
    pub project_id: u32,
    pub task_id: u32,
    pub spent_date: Option<String>,
    pub notes: Option<String>,
    pub is_running: bool,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct ProjectPage {
    pub projects: Vec<Project>,
    pub per_page: u32,
    pub total_pages: u32,
    pub total_entries: u32,
    pub page: u32,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct ProjectAssignmentPage {
    pub project_assignments: Vec<ProjectAssignment>,
    pub per_page: u32,
    pub total_pages: u32,
    pub total_entries: u32,
    pub page: u32,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TimeEntryPage {
    pub time_entries: Vec<TimeEntry>,
    pub per_page: u32,
    pub total_pages: u32,
    pub total_entries: u32,
    pub page: u32,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TaskAssignmentPage {
    pub task_assignments: Vec<TaskAssignment>,
    pub per_page: u32,
    pub total_pages: u32,
    pub total_entries: u32,
    pub page: u32,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct User {
    pub id: u32,
}

impl Project {
    pub fn name_and_code(&self) -> String {
        if self.code == None || self.code.as_ref().unwrap() == "" {
            self.name.clone()
        } else {
            format!("[{}] {}", self.code.as_ref().unwrap(), self.name)
        }
    }
}

impl Harvest {
    const CLIENT_ID: &'static str = "ew1-8t73wKHsqmhRNtxwkBaO";
    const CONFIG_FILE_NAME: &'static str = "timer-for-harvest.json";

    pub fn new() -> Harvest {
        match Harvest::read_authorization_from_file() {
            Some(harvest) => {
                let unix_timestamp = SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .unwrap()
                    .as_secs();
                let one_day = 60 * 60 * 24;

                if harvest.expires_at < unix_timestamp + one_day {
                    Harvest::obtain_new_authorization()
                } else {
                    return harvest;
                }
            }
            None => Harvest::obtain_new_authorization(),
        }
    }

    fn obtain_new_authorization() -> Harvest {
        let listener = TcpListener::bind("127.0.0.1:12345").expect("Port 12345 is already in use");

        Command::new("xdg-open")
            .arg(format!(
                "https://id.getharvest.com/oauth2/authorize?client_id={}&response_type=token",
                Harvest::CLIENT_ID
            ))
            .spawn()
            .expect("Unable to open browser");

        for stream in listener.incoming() {
            let stream = stream.unwrap();
            let result = Harvest::authorize_callback(stream);

            let unix_timestamp = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs();
            let expires_in: u64 = result.2.parse().unwrap();

            let harvest = Harvest {
                token: result.0,
                account_id: result.1.parse().unwrap(),
                expires_at: unix_timestamp + expires_in,
            };
            harvest.write_authorization_to_file();
            return harvest;
        }

        panic!("unable to authorize");
    }

    fn read_authorization_from_file() -> Option<Harvest> {
        match File::open(Harvest::config_file_path()) {
            Ok(mut file) => {
                let mut content = String::new();
                file.read_to_string(&mut content).unwrap();
                Some(
                    serde_json::from_str(&content)
                        .expect(&format!("Invalid configuration file: {}", content).to_string()),
                )
            }
            Err(_) => None,
        }
    }

    fn write_authorization_to_file(&self) {
        write(Harvest::config_file_path(), json!(self).to_string())
            .expect("unable to save config file");
    }

    fn config_file_path() -> PathBuf {
        let mut path = dirs::config_dir().expect("Unable to find XDG config dir path");
        path.push(Harvest::CONFIG_FILE_NAME);
        path
    }

    fn authorize_callback(mut stream: TcpStream) -> (String, String, String) {
        let mut buffer = [0; 512];
        let mut first_line = "".to_string();

        loop {
            match stream.read(&mut buffer) {
                Ok(n) => {
                    if first_line == "" {
                        let request = String::from_utf8_lossy(&buffer[..]).to_string();
                        first_line = request.lines().next().unwrap().to_string();
                    }
                    if n < buffer.len() {
                        break;
                    }
                }
                Err(_) => {
                    panic!("unable to read request");
                }
            }
        }

        let result = parse_account_details(&first_line);

        let response = "HTTP/1.1 200 OK\r\n\r\n<!DOCTYPE html>
            <html>
                <body>
                    Authorized successfully
                </body>
            </html>\r\n";

        stream.write(response.as_bytes()).unwrap();
        stream.flush().unwrap();

        result
    }

    pub fn user_agent() -> String {
        format!(
            "{} {}.{}.{}{} ({})",
            env!("CARGO_PKG_DESCRIPTION"),
            env!("CARGO_PKG_VERSION_MAJOR"),
            env!("CARGO_PKG_VERSION_MINOR"),
            env!("CARGO_PKG_VERSION_PATCH"),
            option_env!("CARGO_PKG_VERSION_PRE").unwrap_or(""),
            env!("CARGO_PKG_HOMEPAGE")
        )
    }

    pub fn active_project_assignments(&self) -> Vec<ProjectAssignment> {
        let mut project_assignments: Vec<ProjectAssignment> = vec![];
        let mut current_page = 1;

        loop {
            let url = format!(
                "https://api.harvestapp.com/v2/users/me/project_assignments?page={}",
                current_page
            );
            let res = self.api_get_request(&url);
            let body = &res.unwrap().text().unwrap();
            let page: ProjectAssignmentPage = serde_json::from_str(body).expect(
                &format!("Unexpected project assignment page structure: {}", body).to_string(),
            );

            for project_assignment in page.project_assignments {
                project_assignments.push(project_assignment);
            }

            if current_page == page.total_pages {
                break;
            } else {
                current_page += 1;
            }
        }

        project_assignments
    }

    pub fn time_entries_for(&self, user: &User, from: String, till: String) -> Vec<TimeEntry> {
        let url = format!(
            "https://api.harvestapp.com/v2/time_entries?user_id={}&from={}&to={}",
            user.id, from, till
        );
        match self.api_get_request(&url) {
            Ok(res) => {
                let body = &res.text().unwrap();
                let page: TimeEntryPage = serde_json::from_str(body)
                    .expect(&format!("Unexpected time entry page structure: {}", body).to_string());

                page.time_entries
            }
            Err(_e) => Vec::new(),
        }
    }

    pub fn current_user(&self) -> User {
        let url = "https://api.harvestapp.com/v2/users/me";
        let res = self.api_get_request(&url);
        let body = &res.unwrap().text().unwrap();
        serde_json::from_str(body)
            .expect(&format!("Unexpected user structure: {}", body).to_string())
    }

    pub fn start_timer(
        &self,
        project_id: u32,
        task_id: u32,
        notes: String,
        hours: f32,
        now: &chrono::NaiveDate,
    ) -> TimeEntry {
        let url = "https://api.harvestapp.com/v2/time_entries";
        let mut timer = Timer {
            id: None,
            project_id: project_id,
            task_id: task_id,
            spent_date: Some(now.format("%Y-%m-%d").to_string()),
            notes: None,
            hours: None,
            is_running: true,
        };
        if notes.len() > 0 {
            timer.notes = Some(notes);
        }
        if hours > 0.0 {
            timer.hours = Some(hours);
        }

        let res = self.api_post_request(&url, &timer);
        let body = &res.text().unwrap();
        serde_json::from_str(body)
            .expect(&format!("Unexpected timer structure: {}", body).to_string())
    }

    pub fn restart_timer(&self, time_entry_id: u32) -> TimeEntry {
        let url = format!(
            "https://api.harvestapp.com/v2/time_entries/{}/restart",
            time_entry_id
        );

        let res = self.api_patch_request(&url, &());
        let body = &res.text().unwrap();
        serde_json::from_str(body)
            .expect(&format!("Unexpected timer structure: {}", body).to_string())
    }

    pub fn stop_timer(&self, time_entry_id: u32) -> TimeEntry {
        let url = format!(
            "https://api.harvestapp.com/v2/time_entries/{}/stop",
            time_entry_id
        );

        let res = self.api_patch_request(&url, &());
        let body = &res.text().unwrap();
        serde_json::from_str(body)
            .expect(&format!("Unexpected timer structure: {}", body).to_string())
    }

    pub fn update_timer(
        &self,
        id: u32,
        project_id: u32,
        task_id: u32,
        notes: String,
        hours: f32,
        is_running: bool,
        spent_date: String,
    ) -> TimeEntry {
        let url = format!("https://api.harvestapp.com/v2/time_entries/{}", id);

        /* TODO how not to sent hours when is_running in a better way? */
        if is_running {
            let t2 = TimerWithoutHours {
                id: Some(id),
                project_id: project_id,
                task_id: task_id,
                notes: Some(notes),
                is_running: is_running,
                spent_date: Some(spent_date),
            };

            let res = self.api_patch_request(&url, &t2);
            let body = &res.text().unwrap();
            serde_json::from_str(body)
                .expect(&format!("Unexpected time entry structure: {}", body).to_string())
        } else {
            let timer = Timer {
                id: Some(id),
                project_id: project_id,
                task_id: task_id,
                notes: Some(notes),
                is_running: is_running,
                hours: Some(hours),
                spent_date: Some(spent_date),
            };
            let res = self.api_patch_request(&url, &timer);
            let body = &res.text().unwrap();
            serde_json::from_str(body)
                .expect(&format!("Unexpected time entry structure: {}", body).to_string())
        }
    }

    pub fn delete_timer(&self, timer_id: u32) -> TimeEntry {
        let url = format!("https://api.harvestapp.com/v2/time_entries/{}", timer_id);

        let res = self.api_delete_request(&url);
        let body = &res.text().unwrap();
        serde_json::from_str(body)
            .expect(&format!("Unexpected timer structure: {}", body).to_string())
    }

    fn api_get_request(&self, url: &str) -> Result<reqwest::blocking::Response, reqwest::Error> {
        let client = reqwest::blocking::Client::new();

        client
            .get(url)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Harvest-Account-Id", format!("{}", self.account_id))
            .header("User-Agent", Harvest::user_agent())
            .send()
    }

    fn api_post_request<T: serde::Serialize + ?Sized>(
        &self,
        url: &str,
        json: &T,
    ) -> reqwest::blocking::Response {
        let client = reqwest::blocking::Client::new();

        client
            .post(url)
            .json(&json)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Harvest-Account-Id", format!("{}", self.account_id))
            .header("User-Agent", Harvest::user_agent())
            .send()
            .unwrap()
    }

    fn api_delete_request(&self, url: &str) -> reqwest::blocking::Response {
        let client = reqwest::blocking::Client::new();

        client
            .delete(url)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Harvest-Account-Id", format!("{}", self.account_id))
            .header("User-Agent", Harvest::user_agent())
            .send()
            .unwrap()
    }

    fn api_patch_request<T: serde::Serialize + ?Sized>(
        &self,
        url: &str,
        json: &T,
    ) -> reqwest::blocking::Response {
        let client = reqwest::blocking::Client::new();

        client
            .patch(url)
            .json(&json)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Harvest-Account-Id", format!("{}", self.account_id))
            .header("User-Agent", Harvest::user_agent())
            .send()
            .unwrap()
    }
}

/* TODO move to TimeEntry */
pub fn duration_str_to_f32(duration: &str) -> f32 {
    if duration.len() > 0 {
        let mut parts = duration.split(":");
        let hours: f32 = match parts.next() {
            None => 0.0,
            Some(h) => match h.parse() {
                Ok(p) => p,
                Err(_) => 0.0,
            },
        };
        let minutes: f32 = match parts.next() {
            None => 0.0,
            Some(m) => match m.parse() {
                Ok(p) => p,
                Err(_) => 0.0,
            },
        };
        hours + minutes / 60.0
    } else {
        0.0
    }
}

/* TODO move to TimeEntry */
pub fn f32_to_duration_str(duration: f32) -> String {
    let mut minutes = duration % 1.0;
    let mut hours = duration - minutes;

    if format!("{:0>2.0}", minutes * 60.0) == "60" {
        minutes = 0.0;
        hours += 1.0;
    }
    format!("{:.0}:{:0>2.0}", hours, minutes * 60.0)
}

/* TODO improve this messy parse function */
pub fn parse_account_details(request: &str) -> (String, String, String) {
    let mut parts = request.split(" ");
    parts.next(); /* GET */
    let uri = parts.next().unwrap().parse::<hyper::Uri>().unwrap();
    let parts = uri.query().unwrap().split("&");
    let mut access_token = "";
    let mut account_id = "";
    let mut expires_in = "";

    for part in parts {
        if part.starts_with("access_token") {
            let mut parts = part.split("=");
            parts.next();
            access_token = parts.next().unwrap();
        } else if part.starts_with("scope") {
            let mut parts = part.split("=");
            parts.next();
            account_id = parts.next().unwrap();
            let mut parts = account_id.split("%3A");
            parts.next();
            account_id = parts.next().unwrap();
        } else if part.starts_with("expires_in") {
            let mut parts = part.split("=");
            parts.next();
            expires_in = parts.next().unwrap();
        }
    }
    (
        access_token.to_string(),
        account_id.to_string(),
        expires_in.to_string(),
    )
}

pub fn escape_html(subject: &str) -> String {
    subject
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace("\"", "&quot;")
        .replace("'", "&#x27;")
        .replace("`", "&#x60;")
}

pub fn format_timeentry_notes_for_list(n: &str, length: Option<usize>) -> std::string::String {
    let take: usize = match length {
        Some(value) => value,
        None => 80,
    };

    let formatted: String = escape_html(n)
        .replace("\n\n", "\n")
        .replace("\n", " - ")
        .chars()
        .take(take)
        .collect();

    if n.chars().count() > take {
        formatted.clone() + "..."
    } else {
        formatted
    }
}
